posts/marker-gene-smoothing/index.qmd

```diff
@@ -0,0 +1,627 @@
+  setnames(umapd, c(names(umapd)[1:3]), c("cell_ID", "UMAP_1", "UMAP_2"))
+  
+  nn_umap <- RANN::nn2(umapd[,.(UMAP_1, UMAP_2)],k = n_neighbors + 1)$nn.idx
+  nn_umap <- data.table::melt(cbind(umapd[,.(cell_ID)], data.table(nn_umap))
+                              , id.vars=c("cell_ID", "V1"))
+  colnames(nn_umap) <- c("cell_ID1", "cell_ID1_idx", "neighbor", "cell_ID2_idx")
+  nn_umap <- merge(nn_umap
+                   , nn_umap[,.(cell_ID2=cell_ID1, cell_ID2_idx=cell_ID1_idx)][
+                     ,unique(.SD)],by="cell_ID2_idx")
+  
+  wumap <- Matrix::sparseMatrix(i = c(unique(nn_umap$cell_ID1_idx), nn_umap$cell_ID2_idx)
+                                ,j=c(unique(nn_umap$cell_ID1_idx), nn_umap$cell_ID1_idx)
+                                ,x=1)
+  
+  rownames(wumap) <- colnames(wumap) <- nn_umap[order(nn_umap$cell_ID1_idx),unique(cell_ID1)]
+  mumap <- Matrix::sparseMatrix(i=1:ncol(wumap)
+                                ,j=1:ncol(wumap)
+                                ,x=1/Matrix::colSums(wumap)
+                                )
+  dimnames(mumap) <- dimnames(wumap)
+  smoother <- mumap %*% wumap
+  smoother <- smoother[,colnames(sem)]
+  return(smoother)
+}
+
+```
+
+Here, we can apply our 'smoothing matrix' to some T cell marker genes and compare the difference on the UMAP.
+Whereas raw marker expression is noisy and sparse, smoothed expression clearly highlights the focal points for each marker on the UMAP.
+
+```{r}
+#| eval: false
+#| code-fold: true
+
+smoother <- umap_nn(sem, "pearsonumap", n_neighbors = 500)
+sem <- SetAssayData(sem, "data", assay = "RNA", new.data = totalcount_norm(sem[["RNA"]]@counts))
+smoothed_tcell_markers <- sem[["RNA"]]@data[c("FOXP3","CD3G","CD8B"),] %*% smoother
+
+plist <- list()
+for(g in c("FOXP3", "CD8B","CD3G")){
+  plist[[paste0(g,".smooth")]] <- 
+    marker_umap_plot(smoothed_tcell_markers, g, sem, "pearsonumap"
+                     ,xlim = c(-5.5, 6.5), ylim = c(-0.01, 7.2)) + 
+        labs(title=paste0("smooth ",g))
+  plist[[paste0(g,".raw")]] <- 
+    marker_umap_plot(sem[["RNA"]]@counts, g, sem, "pearsonumap"
+                     ,xlim = c(-5.5, 6.5), ylim = c(-0.01, 7.2)) + 
+        labs(title=paste0("raw ",g))
+}
+cowplot::plot_grid(plotlist=plist, nrow=3) + 
+  theme(plot.background = element_rect(fill = 'white',color='white')) + 
+  cowplot::panel_border(remove=TRUE) 
+
+```
+
+```{r}
+#| eval: true
+#| echo: false
+
+knitr::include_graphics("./figures/smooth_vs_raw.png")
+```
+
+From here, we could take a number of different approaches to 'clean up' our T cell calls.
+For example, because Tregs are a subset of T CD4 cells, we might consider reassigning them to T CD4 if they have low 'smoothed FOXP3' expression.
+
+Below, we could rank the Treg cells by their smoothed FOXP3 expression, as a way of measuring our confidence in the cell type call.
+
+If we kept all of our Treg cells as Tregs, then 12.7% would be FOXP3+.
+If we keep only the top 100 cells, 42% would be FOXP3+.  
```

Comment:
I keep picturing a flow-like scatterplot of e.g. CD4 and FOXP3. That could be a helpful way to convey how the smoothed data behaves and give people realistic expectations about how clean their new cell typing will be.